File: src/ultra_lean_mcp_proxy/state.py
"""Session/cache/tool-index state for Ultra Lean MCP Proxy."""

from __future__ import annotations

import hashlib
import json
import re
import time
from dataclasses import dataclass
from typing import Any, Optional


def clone_json(value: Any) -> Any:
    """Clone JSON-serializable data via round-trip."""
    return json.loads(json.dumps(value))


def stable_json_dumps(value: Any) -> str:
    """Deterministic JSON serialization used for hashing."""
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def stable_hash(value: Any) -> str:
    text = stable_json_dumps(value)
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def args_hash(arguments: Any) -> str:
    if arguments is None:
        return stable_hash({})
    return stable_hash(arguments)


def is_mutating_tool_name(tool_name: str) -> bool:
    verbs = [
        "create",
        "update",
        "delete",
        "remove",
        "set",
        "write",
        "insert",
        "patch",
        "post",
        "put",
        "merge",
        "upload",
        "commit",
        # Stateful browser/session-like operations that can invalidate read cache.
        "navigate",
        "open",
        "close",
        "click",
        "type",
        "press",
        "select",
        "hover",
        "drag",
        "drop",
        "scroll",
        "evaluate",
        "execute",
        "goto",
        "reload",
        "back",
        "forward",
    ]
    name = tool_name.lower()
    return any(v in name for v in verbs)


def make_cache_key(session_id: str, server_name: str, tool_name: str, arguments: Any) -> str:
    return f"{session_id}:{server_name}:{tool_name}:{args_hash(arguments)}"


@dataclass
class CacheEntry:
    value: Any
    expires_at: float
    created_at: float
    hits: int = 0


@dataclass
class ToolsHashEntry:
    last_hash: Optional[str] = None
    conditional_hits: int = 0
    updated_at: float = 0.0


class ProxyState:
    """In-memory state for one Ultra Lean MCP Proxy process."""

    def __init__(self, max_cache_entries: int = 5000):
        self.max_cache_entries = max(1, max_cache_entries)
        self._cache: dict[str, CacheEntry] = {}
        self._history: dict[str, Any] = {}
        self._tools: list[dict[str, Any]] = []
        self._tools_hash: dict[str, ToolsHashEntry] = {}

    # Cache
    def cache_get(self, key: str) -> Optional[Any]:
        entry = self._cache.get(key)
        if not entry:
            return None
        now = time.time()
        if entry.expires_at < now:
            del self._cache[key]
            return None
        entry.hits += 1
        return clone_json(entry.value)

    def cache_set(self, key: str, value: Any, ttl_seconds: int):
        now = time.time()
        self._cache[key] = CacheEntry(
            value=clone_json(value),
            created_at=now,
            expires_at=now + max(0, ttl_seconds),
            hits=0,
        )
        self._evict_cache_if_needed()

    def cache_invalidate_prefix(self, prefix: str) -> int:
        removed = 0
        for key in list(self._cache.keys()):
            if key.startswith(prefix):
                self._cache.pop(key, None)
                removed += 1
        return removed

    def _evict_cache_if_needed(self):
        if len(self._cache) <= self.max_cache_entries:
            return
        # Evict lowest-hit then oldest.
        ordered = sorted(
            self._cache.items(),
            key=lambda item: (item[1].hits, item[1].created_at),
        )
        overflow = len(self._cache) - self.max_cache_entries
        for key, _ in ordered[:overflow]:
            self._cache.pop(key, None)

    # Delta history
    def history_get(self, key: str) -> Optional[Any]:
        value = self._history.get(key)
        if value is None:
            return None
        return clone_json(value)

    def history_set(self, key: str, value: Any):
        self._history[key] = clone_json(value)
        if len(self._history) > self.max_cache_entries * 2:
            # Soft bound: trim oldest inserted key.
            first_key = next(iter(self._history))
            self._history.pop(first_key, None)

    def history_invalidate_prefix(self, prefix: str) -> int:
        removed = 0
        for key in list(self._history.keys()):
            if key.startswith(prefix):
                self._history.pop(key, None)
                removed += 1
        return removed

    # Tools index
    def set_tools(self, tools: list[dict[str, Any]]):
        self._tools = clone_json(tools or [])

    def get_tools(self) -> list[dict[str, Any]]:
        return clone_json(self._tools)

    def search_tools(
        self,
        query: str,
        top_k: int = 8,
        include_schemas: bool = True,
    ) -> list[dict[str, Any]]:
        if not self._tools:
            return []

        terms = [t for t in re.findall(r"[a-zA-Z0-9_]+", query.lower()) if t]
        ranked = []
        for tool in self._tools:
            name = str(tool.get("name", ""))
            desc = str(tool.get("description", ""))
            schema = tool.get("inputSchema") or tool.get("input_schema") or {}
            props = schema.get("properties", {}) if isinstance(schema, dict) else {}
            param_text = " ".join(str(k) for k in props.keys())
            haystack = f"{name} {desc} {param_text}".lower()

            score = 0.0
            if query.lower() in name.lower():
                score += 4.0
            for term in terms:
                if term in name.lower():
                    score += 2.0
                if term in desc.lower():
                    score += 1.0
                if term in param_text.lower():
                    score += 1.25
                if term in haystack:
                    score += 0.2
            if score <= 0:
                continue
            ranked.append((score, tool))

        if not ranked:
            ranked = [(0.01, tool) for tool in self._tools]

        ranked.sort(key=lambda item: item[0], reverse=True)
        results = []
        for score, tool in ranked[: max(1, top_k)]:
            item = {
                "name": tool.get("name"),
                "score": round(score, 3),
                "description": tool.get("description", ""),
            }
            if include_schemas:
                schema = tool.get("inputSchema") or tool.get("input_schema")
                if schema is not None:
                    item["inputSchema"] = clone_json(schema)
            results.append(item)
        return results

    # tools_hash_sync scope state
    def tools_hash_get(self, key: str) -> Optional[ToolsHashEntry]:
        entry = self._tools_hash.get(key)
        if entry is None:
            return None
        return ToolsHashEntry(
            last_hash=entry.last_hash,
            conditional_hits=entry.conditional_hits,
            updated_at=entry.updated_at,
        )

    def tools_hash_set_last(self, key: str, tools_hash: str):
        now = time.time()
        entry = self._tools_hash.setdefault(key, ToolsHashEntry())
        if entry.last_hash != tools_hash:
            entry.conditional_hits = 0
        entry.last_hash = tools_hash
        entry.updated_at = now

    def tools_hash_record_hit(self, key: str) -> int:
        now = time.time()
        entry = self._tools_hash.setdefault(key, ToolsHashEntry())
        entry.conditional_hits += 1
        entry.updated_at = now
        return entry.conditional_hits

    def tools_hash_reset_hits(self, key: str):
        now = time.time()
        entry = self._tools_hash.setdefault(key, ToolsHashEntry())
        entry.conditional_hits = 0
        entry.updated_at = now

File: npm/src/state.mjs
/**
 * Session/cache/tool-index state for Ultra Lean MCP Proxy.
 */

import { createHash } from 'node:crypto';

function isPlainObject(value) {
  return Object.prototype.toString.call(value) === '[object Object]';
}

export function canonicalize(value) {
  if (Array.isArray(value)) {
    return value.map((item) => canonicalize(item));
  }
  if (isPlainObject(value)) {
    const out = {};
    const keys = Object.keys(value).sort();
    for (const key of keys) {
      out[key] = canonicalize(value[key]);
    }
    return out;
  }
  return value;
}

export function cloneJson(value) {
  if (value === undefined) return undefined;
  return JSON.parse(JSON.stringify(value));
}

export function stableJsonStringify(value) {
  return JSON.stringify(canonicalize(value));
}

export function stableHash(value) {
  return createHash('sha256').update(stableJsonStringify(value), 'utf-8').digest('hex');
}

export function argsHash(argumentsValue) {
  if (argumentsValue === null || argumentsValue === undefined) {
    return stableHash({});
  }
  return stableHash(argumentsValue);
}

const MUTATING_VERBS = [
  'create',
  'update',
  'delete',
  'remove',
  'set',
  'write',
  'insert',
  'patch',
  'post',
  'put',
  'merge',
  'upload',
  'commit',
  'navigate',
  'open',
  'close',
  'click',
  'type',
  'press',
  'select',
  'hover',
  'drag',
  'drop',
  'scroll',
  'evaluate',
  'execute',
  'goto',
  'reload',
  'back',
  'forward',
];

export function isMutatingToolName(toolName) {
  const name = String(toolName || '').toLowerCase();
  return MUTATING_VERBS.some((verb) => name.includes(verb));
}

export function makeCacheKey(sessionId, serverName, toolName, argumentsValue) {
  return `${sessionId}:${serverName}:${toolName}:${argsHash(argumentsValue)}`;
}

function nowSeconds() {
  return Date.now() / 1000;
}

function toolSearchTerms(query) {
  return String(query || '')
    .toLowerCase()
    .match(/[a-zA-Z0-9_]+/g) || [];
}

export class ProxyState {
  constructor(maxCacheEntries = 5000) {
    this.maxCacheEntries = Math.max(1, Number(maxCacheEntries) || 5000);
    this._cache = new Map(); // key -> {value, expiresAt, createdAt, hits}
    this._history = new Map(); // key -> json
    this._tools = [];
    this._toolsHash = new Map(); // key -> {lastHash, conditionalHits, updatedAt}
  }

  cacheGet(key) {
    const entry = this._cache.get(key);
    if (!entry) return null;
    if (entry.expiresAt < nowSeconds()) {
      this._cache.delete(key);
      return null;
    }
    entry.hits += 1;
    return cloneJson(entry.value);
  }

  cacheSet(key, value, ttlSeconds) {
    const now = nowSeconds();
    this._cache.set(key, {
      value: cloneJson(value),
      createdAt: now,
      expiresAt: now + Math.max(0, Number(ttlSeconds) || 0),
      hits: 0,
    });
    this._evictCacheIfNeeded();
  }

  cacheInvalidatePrefix(prefix) {
    let removed = 0;
    for (const key of this._cache.keys()) {
      if (key.startsWith(prefix)) {
        this._cache.delete(key);
        removed += 1;
      }
    }
    return removed;
  }

  _evictCacheIfNeeded() {
    if (this._cache.size <= this.maxCacheEntries) return;
    const ordered = Array.from(this._cache.entries()).sort((a, b) => {
      const ah = a[1].hits - b[1].hits;
      if (ah !== 0) return ah;
      return a[1].createdAt - b[1].createdAt;
    });
    const overflow = this._cache.size - this.maxCacheEntries;
    for (let i = 0; i < overflow; i++) {
      this._cache.delete(ordered[i][0]);
    }
  }

  historyGet(key) {
    if (!this._history.has(key)) return null;
    return cloneJson(this._history.get(key));
  }

  historySet(key, value) {
    this._history.set(key, cloneJson(value));
    if (this._history.size > this.maxCacheEntries * 2) {
      const firstKey = this._history.keys().next().value;
      if (firstKey !== undefined) {
        this._history.delete(firstKey);
      }
    }
  }

  historyInvalidatePrefix(prefix) {
    let removed = 0;
    for (const key of this._history.keys()) {
      if (key.startsWith(prefix)) {
        this._history.delete(key);
        removed += 1;
      }
    }
    return removed;
  }

  setTools(tools) {
    this._tools = cloneJson(Array.isArray(tools) ? tools : []);
  }

  getTools() {
    return cloneJson(this._tools);
  }

  searchTools(query, topK = 8, includeSchemas = true) {
    if (!Array.isArray(this._tools) || this._tools.length === 0) {
      return [];
    }

    const terms = toolSearchTerms(query);
    const queryLower = String(query || '').toLowerCase();
    const ranked = [];

    for (const tool of this._tools) {
      if (!isPlainObject(tool)) continue;
      const name = String(tool.name || '');
      const desc = String(tool.description || '');
      const schema = (isPlainObject(tool.inputSchema) && tool.inputSchema)
        || (isPlainObject(tool.input_schema) && tool.input_schema)
        || {};
      const properties = isPlainObject(schema.properties) ? schema.properties : {};
      const paramText = Object.keys(properties).join(' ');
      const haystack = `${name} ${desc} ${paramText}`.toLowerCase();

      let score = 0;
      if (queryLower && name.toLowerCase().includes(queryLower)) score += 4;
      for (const term of terms) {
        if (name.toLowerCase().includes(term)) score += 2;
        if (desc.toLowerCase().includes(term)) score += 1;
        if (paramText.toLowerCase().includes(term)) score += 1.25;
        if (haystack.includes(term)) score += 0.2;
      }
      if (score <= 0) continue;
      ranked.push([score, tool]);
    }

    const fallbackRanked = ranked.length > 0
      ? ranked
      : this._tools.map((tool) => [0.01, tool]);

    fallbackRanked.sort((a, b) => b[0] - a[0]);
    const limit = Math.max(1, Number(topK) || 8);
    const out = [];
    for (const [score, tool] of fallbackRanked.slice(0, limit)) {
      const item = {
        name: tool.name,
        score: Number(score.toFixed(3)),
        description: tool.description || '',
      };
      if (includeSchemas) {
        const schema = tool.inputSchema || tool.input_schema;
        if (schema !== undefined) {
          item.inputSchema = cloneJson(schema);
        }
      }
      out.push(item);
    }
    return out;
  }

  toolsHashGet(key) {
    const entry = this._toolsHash.get(key);
    if (!entry) return null;
    return {
      lastHash: entry.lastHash ?? null,
      conditionalHits: Number(entry.conditionalHits || 0),
      updatedAt: Number(entry.updatedAt || 0),
    };
  }

  toolsHashSetLast(key, toolsHash) {
    const now = nowSeconds();
    const current = this._toolsHash.get(key) || {
      lastHash: null,
      conditionalHits: 0,
      updatedAt: 0,
    };
    if (current.lastHash !== toolsHash) {
      current.conditionalHits = 0;
    }
    current.lastHash = toolsHash;
    current.updatedAt = now;
    this._toolsHash.set(key, current);
  }

  toolsHashRecordHit(key) {
    const now = nowSeconds();
    const current = this._toolsHash.get(key) || {
      lastHash: null,
      conditionalHits: 0,
      updatedAt: 0,
    };
    current.conditionalHits += 1;
    current.updatedAt = now;
    this._toolsHash.set(key, current);
    return current.conditionalHits;
  }

  toolsHashResetHits(key) {
    const now = nowSeconds();
    const current = this._toolsHash.get(key) || {
      lastHash: null,
      conditionalHits: 0,
      updatedAt: 0,
    };
    current.conditionalHits = 0;
    current.updatedAt = now;
    this._toolsHash.set(key, current);
  }
}

